kvbc/test/sparse_merkle_storage/db_adapter_unit_test.cpp

```diff
@@ -212,34 +223,132 @@ TEST_P(db_adapter_custom_blockchain, add_block_with_non_provable_keys) {
-  std::string key2{"key2"};
-  const Sliver skey1{std::move(key1)};
-  const Sliver skey2{std::move(key2)};
-  concord::kvbc::v2MerkleTree::DBAdapter::NonProvableKeySet non_provable_key_set = {skey1, skey2};
+  const Sliver skey1{"skey1"};
+  const Sliver skey2{"skey2"};
+  const Sliver skey3{"skey3"};
+
+  const Sliver svalue1{"svalue1"};
+  const Sliver svalue2{"svalue2"};
+  const Sliver svalue3{"svalue3"};
+  concord::kvbc::v2MerkleTree::DBAdapter::NonProvableKeySet non_provable_key_set = {skey1, skey2, skey3};
   auto adapter = DBAdapter{GetParam()->newEmptyDb(), true, non_provable_key_set};
+
+  const auto initialTreeVersion = adapter.getStateVersion();
+  const auto initialHash = adapter.getStateHash();
+
   ASSERT_THROW(adapter.getValue(skey1, 4), NotFoundException);
 
-  const auto updates = SetOfKeyValuePairs{{skey1, defaultSliver}, {skey2, defaultSliver}};
-  const auto numBlocks = 4u;
-  for (auto i = 0u; i < numBlocks; ++i) {
-    ASSERT_EQ(adapter.addBlock(updates), i + 1);
+  const std::vector<SetOfKeyValuePairs> updates = {{{skey1, svalue1}}, {{skey2, svalue2}}, {{skey3, svalue3}}};
+  for (size_t i = 0u; i < updates.size(); ++i) {
+    ASSERT_EQ(adapter.addBlock(updates[i]), i + 1);
   }
 
-  for (const auto &k : non_provable_key_set) {
-    const auto [value, blockId] = adapter.getValue(k, numBlocks);
-    ASSERT_EQ(value, defaultSliver);
-    ASSERT_EQ(blockId, numBlocks);
+  // Make sure getValue throws when block id is incorrect
+  for (size_t i = 0u; i < updates.size(); ++i) {
+    ASSERT_THROW(adapter.getValue(updates[i].begin()->first, i), NotFoundException);
+  }
+
+  // Make sure getValue returns correct value when the exact block id is provided
+  for (size_t i = 0u; i < updates.size(); ++i) {
+    const auto &[value, blockId] = adapter.getValue(updates[i].begin()->first, i + 1);
+    ASSERT_EQ(blockId, i + 1);
+    ASSERT_EQ(value, updates[i].begin()->second);
+  }
+
+  // Make sure getValue returns correct value when last reachable block id is provided
+  for (size_t i = 0u; i < updates.size(); ++i) {
+    const auto &[value, blockId] = adapter.getValue(updates[i].begin()->first, adapter.getLastReachableBlockId());
+    ASSERT_EQ(blockId, i + 1);
+    ASSERT_EQ(value, updates[i].begin()->second);
+  }
+
+  ASSERT_THROW(adapter.getValue(Sliver{"non-exisiting-key"}, adapter.getLastReachableBlockId()), NotFoundException);
+
+  // Make sure that blocks with non-provable keys does not affect the Merkle Tree state
+  ASSERT_EQ(initialTreeVersion, adapter.getStateVersion());
+  ASSERT_EQ(initialHash, adapter.getStateHash());
+
+  // Make sure that adding blocks with provable keys affects the Merkle Tree state
+  adapter.addBlock(SetOfKeyValuePairs{{Sliver{"provable-key"}, Sliver{"provable-value"}}});
+  ASSERT_EQ(initialTreeVersion + 1, adapter.getStateVersion());
+  ASSERT_NE(initialHash, adapter.getStateHash());
+}
+
+TEST_P(db_adapter_custom_blockchain, raw_block_with_non_provable_and_provable_keys) {
```

Comment:
Does it make sense to add blocks that include both provable and non-provable keys (in addition to provable-only and non-provable only blocks)?

Reply:
Since we may have blocks with both types of keys, I think it is safer to test it as well.
Mostly for regression testing's sake.